bentoml/_internal/resource.py

```diff
@@ -0,0 +1,279 @@
+from __future__ import annotations
+
+import os
+import re
+import math
+import typing as t
+import logging
+import functools
+from abc import ABCMeta
+from abc import abstractmethod
+
+import psutil
+
+from ..exceptions import BentoMLConfigException
+
+logger = logging.getLogger(__name__)
+
+_RESOURCE_REGISTRY: dict[str, t.Type[Resource[t.Any]]] = {}
+
+T = t.TypeVar("T")
+
+
+def get_resource(resources: dict[str, t.Any], resource_kind: str) -> t.Any:
+    if resource_kind not in _RESOURCE_REGISTRY:
+        raise BentoMLConfigException(f"Unknown resource kind '{resource_kind}'.")
+
+    resource: t.Type[Resource[t.Any]] = _RESOURCE_REGISTRY[resource_kind]
+
+    if resource_kind in resources:
+        if resources[resource_kind] == "system":
+            return resource.from_system()
+        else:
+            return resource.from_spec(resources[resource_kind])
+    else:
+        return None
+
+
+def system_resources() -> dict[str, t.Any]:
+    res: dict[str, t.Any] = {}
+    for resource_kind, resource in _RESOURCE_REGISTRY.items():
+        res[resource_kind] = resource.from_system()
+    return res
+
+
+class ResourceMeta(ABCMeta):
+    def __new__(  # pylint: disable=arguments-differ # we don't want to be generic
+        cls,
+        name: str,
+        bases: tuple[type],
+        attr_dict: dict[t.Any, t.Any],
+        *,
+        resource_id: str,
+    ):
+        res = super().__new__(cls, name, bases, attr_dict)
+        if resource_id != "":
+            # register the class with RESOURCE_REGISTRY
+            assert issubclass(res, Resource)
+            _RESOURCE_REGISTRY[resource_id] = res
+
+        return res  # pylint: disable=useless-super-delegation (???)
+
+
+class Resource(t.Generic[T], metaclass=ResourceMeta, resource_id=""):
+    @classmethod
+    @abstractmethod
+    def from_spec(cls, spec: t.Any) -> T:
+        """
+        Get an instance of this resource from user input. For example, a CPU resource
+        might parse "10m" and return a CPU resource with 0.01 CPUs.
+        """
+
+    @classmethod
+    @abstractmethod
+    def from_system(cls) -> T:
+        """
+        Infer resource value from the system.
+        """
+
+
+class CpuResource(Resource[float], resource_id="cpu"):
+    @classmethod
+    def from_spec(cls, spec: t.Any) -> float:
+        """
+        Convert spec to CpuResource.
+
+        spec can be a float, int or string.
+        - 1.0 -> 1.0
+        - 1 -> 1.0
+        - "1" -> 1.0
+        - "10m" -> 0.01
+        """
+        if not isinstance(spec, (int, float, str)):
+            raise TypeError("cpu must be int, float or str")
+
+        if isinstance(spec, (int, float)):
+            return float(spec)
+
+        milli_match = re.match("([0-9]+)m", spec)
+        if milli_match:
+            return float(milli_match[1]) / 1000.0
+
+        raise BentoMLConfigException(f"Invalid CPU resource limit '{spec}'. ")
```

Comment:
I think the regex does not current handle the case with a non-milli string value.
```
- "1" -> 1.0
```